package/piksi_system_daemon/piksi_system_daemon/src/cellmodem_probe.c

```diff
@@ -0,0 +1,169 @@
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <termios.h>
+#include <unistd.h>
+#include <stdarg.h>
+
+#include <libpiksi/sbp_zmq_pubsub.h>
+#include <libpiksi/logging.h>
+#include <libsbp/logging.h>
+
+#include "cellmodem.h"
+#include "cellmodem_probe.h"
+
+enum {
+  RESPONSE_OK,
+  RESPONSE_ERROR,
+};
+
+static const char *const response_strs[] = {
+  [RESPONSE_OK] = "OK\r\n",
+  [RESPONSE_ERROR] = "ERROR\r\n",
+  NULL
+};
+
+static int modem_read(int fd, char *buf, size_t len, const char *const *responses)
+{
+  int n = 0;
+  fd_set fds;
+  struct timeval timeout = {.tv_sec = 0, .tv_usec = 100000};
+  do {
+    FD_ZERO(&fds);
+    FD_SET(fd, &fds);
+    if (select(fd+1, &fds, NULL, NULL, &timeout) < 0)
+      return -2;
+    if (!FD_ISSET(fd, &fds)) {
+      return -1;
+    }
+    int r = read(fd, buf + n, len - n - 1);
+    if (r < 0) {
+        return -3;
+    }
+    n += r;
+    buf[n] = 0;
+    for(int i = 0; responses[i]; i++) {
+      int rlen = strlen(responses[i]);
+      if ((n >= rlen) && (strcmp(buf + n - rlen, responses[i]) == 0)) {
+        buf[n - rlen] = 0;
+        return i;
+      }
+    }
+  } while (n < len - 1);
+}
+
+#define DELIM "\r\n"
+static int cellmodem_command(int fd, const char *cmd, char *response, size_t len)
+{
+  /* Send command */
+  write(fd, cmd, strlen(cmd));
+  write(fd, "\r", 1);
+
+  /* Read full response */
+  char buf[256] = {0};
+  if (modem_read(fd, buf, sizeof(buf), response_strs) != RESPONSE_OK)
+    return -1;
+
+  /* Check command echo */
+  if (strncmp(buf, cmd, strlen(cmd)) != 0)
+    return -2;
+
+  /* Pull out response */
+  if (response) {
+    /* Strip \r\n junk */
+    int i;
+    for (i = strlen(buf) - 1; (buf[i] == '\r') || (buf[i] == '\n'); i--)
+      buf[i] = 0;
+    for (i = strlen(cmd); (buf[i] == '\r') || (buf[i] == '\n'); i++)
+      ;
+
+    strncpy(response, buf + i, len);
+  }
+
+  return 0;
+}
+
+static void log_info(sbp_zmq_pubsub_ctx_t *pubsub_ctx, const char *fmt, ...)
+{
+  msg_log_t *msg = alloca(256);
+  msg->level = 6;
+
+  va_list ap;
+  va_start(ap, fmt);
+  vsnprintf(msg->text, 255, fmt, ap);
+  va_end(ap);
+
+  sbp_zmq_tx_send(sbp_zmq_pubsub_tx_ctx_get(pubsub_ctx),
+                 SBP_MSG_LOG, sizeof(*msg) + strlen(msg->text), (void*)msg);
+}
+
+enum modem_type cellmodem_probe(const char *dev, sbp_zmq_pubsub_ctx_t *pubsub_ctx)
+{
+  if ((strncmp(dev, "ttyUSB", 6) != 0) && (strncmp(dev, "ttyACM", 6) != 0))
+    return MODEM_TYPE_INVALID;
+
+  piksi_log(LOG_DEBUG, "%s trying /dev/%s", __func__, dev);
+  char fulldev[PATH_MAX];
+  snprintf(fulldev, sizeof(fulldev), "/dev/%s", dev);
+  int fd = open(fulldev, O_RDWR);
+  if (fd < 0) {
+    piksi_log(LOG_WARNING, "%s: Failed to open %s: %s", __func__, fulldev, strerror(errno));
+    return MODEM_TYPE_INVALID;
+  }
+  if (!isatty(fd)) {
+    close(fd);
+    return MODEM_TYPE_INVALID;
+  }
+
+  /* Set to raw mode */
+  struct termios tio = {0};
+  cfmakeraw(&tio);
+  cfsetspeed(&tio, B115200);
```

Comment:
Is there anyway to add a constant for this?  We hard code the speed to 115200 for the pppd config file as well.  If there's every a possibility of it being different, it would be nice to have a constant.

Reply:
These aren't real serial ports, at least not now, so the actual number is unimportant, but it must be something non-zero.  I think it's set to 9600 by default by the tty driver, which would be good enough.  Maybe I should just do `tcgetattr` and set the parameters I care about?  If we ever indent to support modems where this matters it would most likely need to become a setting.

Reply:
I think either way is fine, but comment might be nice for people without a lot of exposure to pseudo-terminals (like me).  Why does it need to be non-zero?